### PR TITLE
Audio PR1 - Decoder Changes for Audio in C++

### PR DESCRIPTION
**Pull request by swetha097**
- This PR contains the decoder changes for audio pipeline in the c++ end.
- A unit test has also been included in here.

### CODE REVIEW

File: utilities/rocAL/rocAL_audio_unittests/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+### build
+  ````
+  mkdir build
+  cd build
+  cmake ../
+  make
+  ````
+### running the application
+  ````
+Go to MIVisionX-tests/rocal-unittests
```
Comment (swetha097) [inline utilities/rocAL/rocAL_audio_unittests/README.md:21]: Change the names here

File: rocAL/rocAL/include/readers/image/file_source_reader.h [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright (c) 2019 - 2023 Advanced Micro Devices, Inc. All rights reserved.
+Copyright (c) 2019 - 2022 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (swetha097) [inline rocAL/rocAL/include/readers/image/file_source_reader.h:2]: Change this back to 2023

File: rocAL/rocAL/include/readers/image/reader.h [hunk 3]
```diff
@@ -159,6 +159,6 @@ class Reader {
     virtual std::string id() = 0;
     //! Returns the number of items remained in this resource
     virtual unsigned count_items() = 0;
-    
+    virtual std::string file_path() = 0;
```
Comment (swetha097) [inline rocAL/rocAL/include/readers/image/reader.h:162]: This has been introduced for audio source evaluation & audio_read_and_decode where we take the file path instead of the reader->id()

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 4]
```diff
@@ -2534,3 +2552,143 @@ rocalResetLoaders(RocalContext p_context)
...
+        unsigned shard_id,
+        unsigned shard_count,
+        bool is_output,
+        bool shuffle,
+        bool loop,
+        bool downmix,
+        unsigned max_frames,
+        unsigned max_channels,
+        unsigned storage_type,
+        bool stick_to_shard,
```
Comment (swetha097) [inline rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:2570]: Currently stick_to_shard & shard_size is not in use - To be removed.
Will be added as a part of the last batch policy

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 5]
```diff
@@ -2534,3 +2552,143 @@ rocalResetLoaders(RocalContext p_context)
...
+        std::cerr << e.what() << '\n';
+    }
+    return output;
+}
+
+RocalTensor  ROCAL_API_CALL
+rocalAudioFileSource(
+        RocalContext p_context,
+        const char* source_path,
+        const char* source_file_list_path,
```
Comment (swetha097) [inline rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:2624]: To remove this file_list_path now - will be added as a part of the file list reader (meta-data reader)

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 6]
```diff
@@ -2534,3 +2552,143 @@ rocalResetLoaders(RocalContext p_context)
     }
     return ROCAL_OK;
 }
+
+RocalTensor  ROCAL_API_CALL
+rocalAudioFileSourceSingleShard(
+        RocalContext p_context,
+        const char* source_path,
+        const char* source_file_list_path,
```
Comment (swetha097) [inline rocAL/rocAL/source/api/rocal_api_data_loaders.cpp:2560]: To remove this file_list_path now - will be added as a part of the file list reader (meta-data reader)

File: utilities/rocAL/rocAL_audio_unittests/rocAL_audio_unittests.cpp [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    /*>>>>>>>>>>>>>>>>>>> Diplay using OpenCV <<<<<<<<<<<<<<<<<*/
+    const unsigned number_of_cols = 1; //1920 / w;
+    cv::Mat mat_output, mat_input;
+    RocalTensorOutputType tensorOutputType = RocalTensorOutputType::ROCAL_FP32;
+
+    cv::Mat mat_color;
+    int col_counter = 0;
+    //cv::namedWindow("output", CV_WINDOW_AUTOSIZE);
```
Comment (swetha097) [inline utilities/rocAL/rocAL_audio_unittests/rocAL_audio_unittests.cpp:237]: Remove this if not required

File: utilities/rocAL/rocAL_audio_unittests/rocAL_audio_unittests.cpp [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    /*>>>>>>>>>>>>>>>> Creating Rocal parameters  <<<<<<<<<<<<<<<<*/
+
+    rocalSetSeed(0);
+
+
+    RocalMetaData metadata_output;
+    // MetaData reader for input file_list which has file seperated by labels
+    // metadata_output = rocalCreateCOCOReader(handle, json_path, true, false);
```
Comment (swetha097) [inline utilities/rocAL/rocAL_audio_unittests/rocAL_audio_unittests.cpp:117]: Remove the call to coco-reader

File: rocAL/rocAL/source/readers/image/file_source_reader.cpp [hunk 9]
```diff
@@ -156,46 +157,47 @@ void FileSourceReader::reset()
...
-        if(filesys::exists(pathObj) && filesys::is_regular_file(pathObj))
-        {
-            // ignore files with non-image extensions
-            auto file_extension_idx = subfolder_path.find_last_of(".");
-            if (file_extension_idx  != std::string::npos) {
-                std::string file_extension = subfolder_path.substr(file_extension_idx+1);
-                std::transform(file_extension.begin(), file_extension.end(), file_extension.begin(),
-                    [](unsigned char c){ return std::tolower(c); });
-                if ((file_extension != "jpg") && (file_extension != "jpeg") && (file_extension != "png") && (file_extension != "ppm") && (file_extension != "bmp") && (file_extension != "pgm") && (file_extension != "tif") && (file_extension != "tiff") && (file_extension != "webp"))
+    for (auto& entry : filesys::recursive_directory_iterator(_folder_path.c_str())) {
```
Comment (swetha097) [inline rocAL/rocAL/source/readers/image/file_source_reader.cpp:162]: These changes have been done to accommodate the file source reader for structures such as

Recursive folder directories
folder/
folder1/folder1_1/ folder1_2/
folder2/folder2_1/ folder2_2

File: rocAL/rocAL/include/api/rocal_api_augmentation.h [hunk 10]
```diff
@@ -57,8 +57,8 @@ extern "C" RocalTensor  ROCAL_API_CALL rocalResize(RocalContext context, RocalTe
                                                    unsigned resize_shorter = 0,
                                                    unsigned resize_longer = 0,
                                                    RocalResizeInterpolationType interpolation_type = ROCAL_LINEAR_INTERPOLATION,
-                                                   RocalTensorLayout rocal_tensor_output_layout = ROCAL_NHWC,
-                                                   RocalTensorOutputType rocal_tensor_output_type = ROCAL_UINT8);
+                                                   RocalTensorLayout output_layout = ROCAL_NONE,
+                                                   RocalTensorOutputType output_datatype = ROCAL_UINT8);
```
Comment (sampath1117) [inline rocAL/rocAL/include/api/rocal_api_augmentation.h:61]: Are these changes expected?

File: rocAL/rocAL/include/api/rocal_api_data_loaders.h [hunk 11]
```diff
@@ -747,5 +747,61 @@ extern "C"  RocalTensor  ROCAL_API_CALL rocalJpegCaffe2LMDBRecordSourcePartialSi
...
+/// If the files are not in standard audio compression formats they will be ignored.
+/// \param context Rocal context
+/// \param source_path A NULL terminated char string pointing to the location on the disk
+/// \param source_file_list_path A NULL terminated char string pointing to the file list location on the disk
+/// \param shard_count Defines the parallelism level by internally sharding the input dataset and load/decode using multiple decoder/loader instances. Using shard counts bigger than 1 improves the load/decode performance if compute resources (CPU cores) are available.
+/// \param is_output Determines if the user wants the loaded audio to be part of the output or not.
+/// \param shuffle Determines if the user wants to shuffle the dataset or not.
+/// \param loop Determines if the user wants to indefinitely loops through audio or not.
+/// \param downmix If set to True, downmix all input channels to mono. If downmixing is turned on, the decoder output is 1D. If downmixing is turned off, it produces 2D output with interleaved channels.
```
Comment (sampath1117) [inline rocAL/rocAL/include/api/rocal_api_data_loaders.h:759]: you can change this as
"If downmixing is turned off, it produces 2D output with interleaved channels incase of multichannel audio"

File: rocAL/rocAL/include/decoders/audio/audio_decoder.h [hunk 12]
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright (c) 2019 - 2023 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (sampath1117) [inline rocAL/rocAL/include/decoders/audio/audio_decoder.h:2]: Copy right in first few files have been changed from 2019 - 2023 -> 2023
Please check and change the same if that is applicable here

File: rocAL/rocAL/include/decoders/audio/audio_decoder.h [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+#pragma once
+
+#include <cstddef>
+#include <iostream>
+#include <vector>
+#include "parameter_factory.h"
+#include "sndfile.h"
+
+
```
Comment (sampath1117) [inline rocAL/rocAL/include/decoders/audio/audio_decoder.h:31]: Please remove 1 empty blank line

File: rocAL/rocAL/include/decoders/audio/sndfile_decoder.h [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#pragma once
+
+#include "audio_decoder.h"
+
+class SndFileDecoder : public AudioDecoder
+{
```
Comment (sampath1117) [inline rocAL/rocAL/include/decoders/audio/sndfile_decoder.h:28]: Please change braces to google style

File: rocAL/rocAL/include/loaders/audio/audio_loader.h [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+#pragma once
+
+#include <string>
+#include <thread>
+#include <vector>
+#include "commons.h"
+#include "circular_buffer.h"
+#include "audio_read_and_decode.h"
+#include "meta_data_reader.h"
+//
```
Comment (sampath1117) [inline rocAL/rocAL/include/loaders/audio/audio_loader.h:32]: Please remove the additional //

File: rocAL/rocAL/include/loaders/audio/audio_loader.h [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    bool _loop;//<! If true the reader will wrap around at the end of the media (files/audios/...) and wouldn't stop
+    size_t _prefetch_queue_depth; // Used for circular buffer's internal buffer
+    size_t _audio_counter = 0;//!< How many audios have been loaded already
+    size_t _remaining_audio_count;//!< How many audios are there yet to be loaded
+    bool _decoder_keep_original = false;
+    int _device_id;
+    size_t _max_decoded_samples, _max_decoded_channels;
+};
+
+
```
Comment (sampath1117) [inline rocAL/rocAL/include/loaders/audio/audio_loader.h:84]: Please remove empty blank line

File: rocAL/rocAL/include/loaders/audio/audio_read_and_decode.h [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+#include "commons.h"
+#include "sndfile_decoder.h"
+#include "reader_factory.h"
+#include "timing_debug.h"
+#include "loader_module.h"
+#include "audio_decoder.h"
+
+class AudioReadAndDecode
+{
+public:
```
Comment (sampath1117) [inline rocAL/rocAL/include/loaders/audio/audio_read_and_decode.h:35]: Please change braces to google style
Applicable to all such instances

File: rocAL/rocAL/include/loaders/audio/audio_read_and_decode.h [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    std::vector<float*> _decompressed_buff_ptrs;
+    std::vector<size_t> _actual_decoded_samples;
+    std::vector<size_t> _actual_decoded_channels;
+    std::vector<size_t> _original_samples;
+    std::vector<size_t> _original_channels;
+    std::vector<float> _original_sample_rates;
+    TimingDBG _file_load_time, _decode_time;
+    size_t _batch_size, _num_threads;
+    DecoderConfig _decoder_config;
+    std::string _input_path;
```
Comment (sampath1117) [inline rocAL/rocAL/include/loaders/audio/audio_read_and_decode.h:73]: Please add a blank line at EOF

File: rocAL/rocAL/include/loaders/audio/node_audio_loader_single_shard.h [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+*/
+
+#pragma once
+#include "node.h"
+#include "audio_loader_sharded.h"
+#include "graph.h"
+
+class AudioLoaderSingleShardNode: public Node {
+public:
+
```
Comment (sampath1117) [inline rocAL/rocAL/include/loaders/audio/node_audio_loader_single_shard.h:30]: Please remove empty blank line

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 20]
```diff
@@ -35,6 +35,10 @@ THE SOFTWARE.
 #include "node_fused_jpeg_crop_single_shard.h"
 #include "node_video_loader.h"
 #include "node_video_loader_single_shard.h"
+#include "node_audio_loader.h"
+#include "node_audio_loader_single_shard.h"
+#include "node_audio_loader.h"
+#include "node_audio_loader_single_shard.h"
```
Comment (sampath1117) [inline rocAL/rocAL/include/pipeline/master_graph.h:41]: Why includes are added 2 times?

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 21]
```diff
@@ -374,3 +378,34 @@ template<> inline std::shared_ptr<VideoLoaderSingleShardNode> MasterGraph::add_n
 
     return node;
 }
+
+/*
+ * Explicit specialization for AudioLoaderNode
+ */
+template<> inline std::shared_ptr<AudioLoaderNode> MasterGraph::add_node(const std::vector<Tensor*>& inputs, const std::vector<Tensor*>& outputs)
+{
```
Comment (sampath1117) [inline rocAL/rocAL/include/pipeline/master_graph.h:386]: Please change braces to google style
Applicable all such instances in this PR

File: rocAL/rocAL/include/pipeline/tensor.h [hunk 22]
```diff
@@ -186,6 +186,14 @@ class TensorInfo {
     bool is_image() const { return _is_image; }
     void set_metadata() { _is_metadata = true; }
     bool is_metadata() const { return _is_metadata; }
+    void set_roi_ptr(unsigned *roi_ptr) { 
+        auto deleter = [&](unsigned *ptr) {};   // Empty destructor used, since memory is handled by the pipeline
+        _roi.reset(roi_ptr, deleter); 
+    }
+    void copy_roi(void *roi_buffer) {
+        if(_roi != nullptr && roi_buffer != nullptr)
+            memcpy((void *)roi_buffer, (const void *)_roi.get(), _batch_size * sizeof(RocalROI));
```
Comment (sampath1117) [inline rocAL/rocAL/include/pipeline/tensor.h:195]: Why are we redoing cast to void* for roi buffer?
use static cast instead of c style casting wherever needed

File: rocAL/rocAL/include/readers/image/caffe2_lmdb_record_reader.h [hunk 23]
```diff
@@ -60,6 +60,8 @@ class Caffe2LMDBRecordReader : public Reader
     //! Returns the id of the latest file opened
     std::string id() override { return _last_id;};
 
+    std::string file_path() override { return _last_file_name; }
```
Comment (sampath1117) [inline rocAL/rocAL/include/readers/image/caffe2_lmdb_record_reader.h:63]: Is this change needed in this file?

**Review by sampath1117 — COMMENTED** [review]
@swetha097 
Added 1st round of review comments
Please address them

File: rocAL/rocAL/source/pipeline/tensor.cpp [hunk 24]
```diff
@@ -63,6 +63,8 @@ vx_size tensor_data_size(RocalTensorDataType data_type) {
             return sizeof(vx_uint32);
         case RocalTensorDataType::INT32:
             return sizeof(vx_int32);
+        case RocalTensorDataType::INT8:
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/pipeline/tensor.cpp:66]: Has int8 been tested with pybind for audio pipeline?

File: rocAL/rocAL/source/pipeline/tensor.cpp [hunk 25]
```diff
@@ -77,6 +79,12 @@ vx_enum interpret_tensor_data_type(RocalTensorDataType data_type) {
             return VX_TYPE_FLOAT16;
         case RocalTensorDataType::UINT8:
             return VX_TYPE_UINT8;
+        case RocalTensorDataType::INT8:
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/pipeline/tensor.cpp:82]: Are you adding these types here to support audio of all dtypes? If so, we need to make sure pybind supports tensors of these types first

File: rocAL/rocAL/source/readers/image/file_source_reader.cpp [hunk 26]
```diff
@@ -156,46 +157,47 @@ void FileSourceReader::reset()
...
+            auto last_slash_idx = entry_path_id.find_last_of("\\/");
+            if (std::string::npos != last_slash_idx)
+            {
+                entry_path_id.erase(0, last_slash_idx + 1);
+            }
+            if (filesys::is_regular_file(entry.path() ))
+            {
+                if(!_meta_data_reader || _meta_data_reader->exists(entry_path_id))
+                {
+                if(get_file_shard_id() != _shard_id )
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/readers/image/file_source_reader.cpp:175]: Indent the if block

File: rocAL/rocAL/source/readers/image/file_source_reader.cpp [hunk 27]
```diff
@@ -156,46 +157,47 @@ void FileSourceReader::reset()
...
+                {
+                    _file_count_all_shards++;
+                    incremenet_file_id();
                     continue;
+                }
+                _in_batch_read_count++;
+                _in_batch_read_count = (_in_batch_read_count%_batch_count == 0) ? 0 : _in_batch_read_count;
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/readers/image/file_source_reader.cpp:182]: Ensure all operations has space before and after

File: rocAL/rocAL/source/readers/image/file_source_reader.cpp [hunk 28]
```diff
@@ -251,14 +253,13 @@ Reader::Status FileSourceReader::open_folder()
...
         file_path.append(_entity->d_name);
+        _last_file_name = file_path;
         _file_names.push_back(file_path);
         _file_count_all_shards++;
         incremenet_file_id();
     }
     if(_file_names.empty())
         WRN("FileReader ShardID ["+ TOSTR(_shard_id)+ "] Did not load any file from " + _folder_path)
-    std::sort(_file_names.begin(), _file_names.end());
-    _last_file_name = _file_names[_file_names.size()-1];
```
Comment (SundarRajan28) [inline rocAL/rocAL/source/readers/image/file_source_reader.cpp:263]: Why have u removed the sorting of filenames?

**Review by SundarRajan28 — COMMENTED** [review]
I have reviewed all the existing file changes for this PR. I will review the new file additions next.

File: utilities/rocAL/rocAL_audio_unittests/README.md (hunk 1 above)
Comment (swetha097) [reply utilities/rocAL/rocAL_audio_unittests/README.md:21]: done

File: rocAL/rocAL/include/readers/image/file_source_reader.h (hunk 2 above)
Comment (swetha097) [reply rocAL/rocAL/include/readers/image/file_source_reader.h]: done

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp (hunk 4 above)
Comment (swetha097) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp]: done

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp (hunk 5 above)
Comment (swetha097) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp]: done

File: utilities/rocAL/rocAL_audio_unittests/rocAL_audio_unittests.cpp (hunk 7 above)
Comment (swetha097) [reply utilities/rocAL/rocAL_audio_unittests/rocAL_audio_unittests.cpp]: done

File: utilities/rocAL/rocAL_audio_unittests/rocAL_audio_unittests.cpp (hunk 8 above)
Comment (swetha097) [reply utilities/rocAL/rocAL_audio_unittests/rocAL_audio_unittests.cpp]: done

File: rocAL/rocAL/include/api/rocal_api_data_loaders.h (hunk 11 above)
Comment (swetha097) [reply rocAL/rocAL/include/api/rocal_api_data_loaders.h]: done

File: amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp [hunk 29]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright (c) 2019 - 2023 Advanced Micro Devices, Inc. All rights reserved.
+Copyright (c) 2023 Advanced Micro Devices, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (swetha097) [reply amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp]: This change is not a part of this PR, not sure why it shows up here. Need to investigate

File: rocAL/rocAL/include/api/rocal_api_augmentation.h (hunk 10 above)
Comment (swetha097) [reply rocAL/rocAL/include/api/rocal_api_augmentation.h]: This change is not a part of this PR, not sure why it shows up here. Need to investigate

File: rocAL/rocAL/include/decoders/audio/audio_decoder.h (hunk 12 above)
Comment (swetha097) [reply rocAL/rocAL/include/decoders/audio/audio_decoder.h:2]: Those changes are not a part of this PR, Will check with fiona

File: rocAL/rocAL/include/loaders/audio/audio_read_and_decode.h (hunk 18 above)
Comment (swetha097) [reply rocAL/rocAL/include/loaders/audio/audio_read_and_decode.h:73]: It shows thats it already added to me

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 20 above)
Comment (swetha097) [reply rocAL/rocAL/include/pipeline/master_graph.h:41]: I guess during some merge, this would have been changes. Will revert it

File: rocAL/rocAL/include/pipeline/tensor.h (hunk 22 above)
Comment (swetha097) [reply rocAL/rocAL/include/pipeline/tensor.h:195]: This changes will be a part of @fiona-gladwin's ROI PR.

File: rocAL/rocAL/include/loaders/circular_buffer.h [hunk 30]
```diff
@@ -30,13 +30,16 @@ THE SOFTWARE.
 #include "device_manager.h"
 #include "device_manager_hip.h"
 #include "commons.h"
-struct decoded_image_info
+struct decoded_sample_info
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/loaders/circular_buffer.h:33]: This is common for all type of data rocAL can handle.
Hence a common name thats suitable for Audio / Image / Video data is sample.

Audio sample/ Image sample/ Video Sample

File: rocAL/rocAL/include/pipeline/tensor.h [hunk 31]
```diff
@@ -210,6 +212,8 @@ class TensorInfo {
     uint64_t _data_size = 0;
     std::vector<size_t> _max_shape;  //!< stores the the width and height dimensions in the tensor
     void reset_tensor_roi_buffers();
+    std::shared_ptr<std::vector<float>> _sample_rate;
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/pipeline/tensor.h:215]: It is needed for the Resample augmentation Node, hence needs to be a part of the TensorInfo

File: rocAL/rocAL/source/pipeline/tensor.cpp (hunk 24 above)
Comment (swetha097) [reply rocAL/rocAL/source/pipeline/tensor.cpp:66]: We only use fp32 dtype for audio for now, so no these other dtypes have not been tested with audio pipeline

File: rocAL/rocAL/source/pipeline/tensor.cpp (hunk 25 above)
Comment (swetha097) [reply rocAL/rocAL/source/pipeline/tensor.cpp:82]: These are added for future use

File: rocAL/rocAL/include/readers/image/file_source_reader.h [hunk 32]
```diff
@@ -55,6 +55,9 @@ class FileSourceReader : public Reader {
     //! Returns the name of the latest file opened
     std::string id() override { return _last_id;};
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/readers/image/file_source_reader.h:57]: @shobana-mcw
Can this be a pure n virtual function in reader.h?

for the other readers, I havent introduced _last_file_path varibale in the open() function , have used _last_file_name which was pre-existing.

File: rocAL/rocAL/source/readers/image/file_source_reader.cpp (hunk 26 above)
Comment (swetha097) [reply rocAL/rocAL/source/readers/image/file_source_reader.cpp]: done

File: amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (swetha097) [reply amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp]: Rebased and this issue is resolved

File: rocAL/rocAL/include/decoders/audio/audio_decoder.h (hunk 12 above)
Comment (swetha097) [reply rocAL/rocAL/include/decoders/audio/audio_decoder.h:2]: Rebased and this issue is resolved

File: rocAL/rocAL/include/loaders/audio/audio_read_and_decode.h (hunk 17 above)
Comment (swetha097) [reply rocAL/rocAL/include/loaders/audio/audio_read_and_decode.h:34]: Done

File: rocAL/rocAL/source/readers/image/file_source_reader.cpp (hunk 27 above)
Comment (swetha097) [reply rocAL/rocAL/source/readers/image/file_source_reader.cpp]: done

File: rocAL/rocAL/include/meta_data/meta_data_graph.h [hunk 33]
```diff
@@ -34,8 +34,8 @@ class MetaDataGraph
 public:
     virtual ~MetaDataGraph()= default;
     virtual void process(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data) = 0;
-    virtual void update_meta_data(pMetaDataBatch meta_data, decoded_image_info decoded_image_info) = 0;
-    virtual void update_random_bbox_meta_data(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data, decoded_image_info decoded_image_info,crop_image_info crop_image_info) = 0;
+    virtual void update_meta_data(pMetaDataBatch meta_data, decoded_sample_info decoded_image_info) = 0;
+    virtual void update_random_bbox_meta_data(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data, decoded_sample_info decoded_image_info,crop_image_info crop_image_info) = 0;
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/meta_data/meta_data_graph.h]: done

File: rocAL/rocAL/include/pipeline/commons.h [hunk 34]
```diff
@@ -132,4 +132,8 @@ struct Timing
     long long unsigned video_read_time= 0;
     long long unsigned video_decode_time= 0;
     long long unsigned video_process_time= 0;
+
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/pipeline/commons.h]: done

File: rocAL/rocAL/include/readers/image/caffe2_lmdb_record_reader.h [hunk 35]
```diff
@@ -81,8 +83,7 @@ class Caffe2LMDBRecordReader : public Reader
     std::map<std::string, unsigned int > _file_size;
     unsigned  _curr_file_idx;
     unsigned _current_file_size;
-    std::string _last_id;
-    std::string _last_file_name;
+    std::string _last_id, _last_file_name;
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/readers/image/caffe2_lmdb_record_reader.h]: done

File: rocAL/rocAL/include/readers/image/caffe_lmdb_record_reader.h [hunk 36]
```diff
@@ -78,8 +80,7 @@ class CaffeLMDBRecordReader : public Reader{
     std::map<std::string, unsigned int > _file_size;
     unsigned  _curr_file_idx;
     unsigned _current_file_size;
-    std::string _last_id;
-    std::string _last_file_name;
+    std::string _last_id, _last_file_name;
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/readers/image/caffe_lmdb_record_reader.h]: done

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 37]
```diff
@@ -2534,3 +2552,143 @@ rocalResetLoaders(RocalContext p_context)
...
+        unsigned max_channels,
+        unsigned storage_type,
+        bool stick_to_shard,
+        int shard_size) {
+    Tensor* output = nullptr;
+    auto context = static_cast<Context*>(p_context);
+    try {
+        if(shard_count < 1 )
+            THROW("Shard count should be bigger than 0")
+
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp]: done

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 38]
```diff
@@ -2534,3 +2552,143 @@ rocalResetLoaders(RocalContext p_context)
...
+        if(shard_id >= shard_count)
+            THROW("Shard id should be smaller than shard count")
+        auto [max_frames, max_channels] = evaluate_audio_data_set(StorageType::FILE_SYSTEM, DecoderType::SNDFILE, source_path, "");
+        INFO("Internal buffer size for audio frames = "+ TOSTR(max_frames))
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp]: done

File: rocAL/rocAL/source/api/rocal_api_data_loaders.cpp [hunk 39]
```diff
@@ -2534,3 +2552,143 @@ rocalResetLoaders(RocalContext p_context)
...
+        bool downmix,
+        unsigned max_frames,
+        unsigned max_channels,
+        unsigned storage_type,
+        bool stick_to_shard,
+        int shard_size) {
+    Tensor* output = nullptr;
+    auto context = static_cast<Context*>(p_context);
+    try {
+        if(shard_count < 1 )
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/source/api/rocal_api_data_loaders.cpp]: done

File: rocAL/rocAL/include/meta_data/bounding_box_graph.h [hunk 40]
```diff
@@ -33,8 +33,8 @@ class BoundingBoxGraph : public MetaDataGraph
 {
 public:
     void process(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data) override;
-    void update_meta_data(pMetaDataBatch meta_data, decoded_image_info decode_image_info) override;
-    void update_random_bbox_meta_data(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data, decoded_image_info decoded_image_info,crop_image_info crop_image_info) override;
+    void update_meta_data(pMetaDataBatch meta_data, decoded_sample_info decode_image_info) override;
+    void update_random_bbox_meta_data(pMetaDataBatch input_meta_data, pMetaDataBatch output_meta_data, decoded_sample_info decoded_image_info,crop_image_info crop_image_info) override;
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/meta_data/bounding_box_graph.h]: done

File: rocAL/rocAL/include/decoders/audio/audio_decoder.h (hunk 13 above)
Comment (swetha097) [reply rocAL/rocAL/include/decoders/audio/audio_decoder.h]: done

File: rocAL/rocAL/include/loaders/audio/audio_loader.h (hunk 16 above)
Comment (swetha097) [reply rocAL/rocAL/include/loaders/audio/audio_loader.h]: done

File: rocAL/rocAL/include/loaders/audio/node_audio_loader_single_shard.h (hunk 19 above)
Comment (swetha097) [reply rocAL/rocAL/include/loaders/audio/node_audio_loader_single_shard.h]: done

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 21 above)
Comment (swetha097) [reply rocAL/rocAL/include/pipeline/master_graph.h]: done

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 20 above)
Comment (swetha097) [reply rocAL/rocAL/include/pipeline/master_graph.h]: reverted
